deep_qa/data/fields/field.py

```diff
@@ -0,0 +1,117 @@
+actually instantiate any PaddedField objects in this code; ``IndexedFields`` just return arrays
+after padding.  We just mention this here as the final endpoint of ``RawFields``.
+"""
+from typing import Dict, List
+
+import numpy
+
+from ..vocabulary import Vocabulary
+
+# Because of how the dependencies work between these objects, we're defining them in reverse order,
+# so that TokenizedField can reference IndexedField in its return types, and RawField can reference
+# TokenizedField.
+
+class IndexedField:
+    """
+    An ``IndexedField`` is a field that is almost ready to be put into a data array for use in a
+    model.  The only thing missing is padding.  The methods on this object allow you to get padding
+    lengths, so you can group things into batches of similar size, and then actually pad the
+    field into a fixed-size array that can be used in a batch of instances.
+    """
+    def get_padding_lengths(self) -> Dict[str, int]:
+        """
+        If there are things in this field that need padding, note them here.  In order to pad a
+        batch of instance, we get all of the lengths from the batch, take the max, and pad
+        everything to that length (or use a pre-specified maximum length).  The return value is a
+        dictionary mapping keys to lengths, like {'num_tokens': 13}.
+        """
+        raise NotImplementedError
+
+    def pad(self, padding_lengths: Dict[str, int]) -> List[numpy.array]:
+        """
+        Given a set of specified padding lengths, actually pad the data in this field and return a
+        numpy array of the correct shape.  This actually returns a list instead of a single array,
+        in case there are several related arrays for this field (e.g., a ``TextField`` might have a
+        word array and a characters-per-word array).
+        """
+        raise NotImplementedError
+
+
+class TokenizedField:
+    """
+    A ``TokenizedField`` is a field that has its raw strings broken up into tokens.  It is used for
+    computing vocabularies.  After computing a vocabulary, we convert ``TokenizedFields`` into
+    ``IndexedFields`` using the vocabulary.
+    """
+    def count_vocab_items(self, vocab: Vocabulary):
+        """
+        If there are strings in this field that need to be converted into integers through a
+        :class:`Vocabulary`, here is where we count them, to determine which tokens are in or out
+        of the vocabulary.
+
+        We pass here instead of raising ``NotImplementedError``, because it is not expected that
+        all fields will have vocabulary items to count.
+        """
+        pass
+
+    def index(self, vocab: Vocabulary) -> IndexedField:
+        """
+        Given a :class:`Vocabulary`, converts all tokens in this field into (typically) integer
+        arrays, returning an ``IndexedField`` as a result.
+        """
+        raise NotImplementedError
+
+
+class RawField:
```

Comment:
If we're going to allow `Fields` to be created at any level of this pipeline, I think we should stick with `Raw` here (or maybe `Untokenized`, or similar).